src/Castle.Core/DynamicProxy/Generators/GeneratorUtil.cs

```diff
@@ -40,7 +40,10 @@ public static class GeneratorUtil
 					continue;
 				}
 
-				emitter.CodeBuilder.AddStatement(AssignArgument(dereferencedArguments, i, arguments));
+				emitter.CodeBuilder.AddStatement(
+					new IfStatement(
+						condition: new MethodInvocationExpression(invocation, InvocationMethods.CouldArgumentValueHaveChanged, new LiteralIntExpression(i)),
+						then: AssignArgument(dereferencedArguments, i, arguments)));
```

Comment:
This is the important bit. Writing back arguments to the `ref` or `out` parameters is skipped in cases where their values still match the ones originally passed in (and stored in `originalArguments`).